### PR TITLE
Adding HLD of ADDB Monitoring

### PR DESCRIPTION
**Pull request by VenkyOS**
Signed-off-by: Venkataraman Padmanabhan <venkataraman.padmanabhan@seagate.com>

### CODE REVIEW

File: doc/ADDB_Monitoring.rst [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+ Execution of monitor
+======================
+
+Partial execution i.e. top half is executed for all the addb records during monitor invocation. This just updates the monitor specific summary data or posts an ast if an exception had been encountered.
+
+There is a periodic posting of these addb summary records and this is done by the locality thread on the m0d. This thread also runs the ASTs that are posted for the exception summary records. 
+
+The bottom half i.e. AST part would be run by a dedicated thread & would be synchronized among the various others threads that would run monitors with a sm (state machine) group lock.
+
+ Conformance
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:146]: "Conformance" has the same level with "Functional Specification" and "Logical Specification".

File: doc/ADDB_Monitoring.rst [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+- [r.addb.monitor.nesting] Monitors generate addb records which themselves can be monitored. 
+
+- [r.addb.stats-service.state] Implementation of stats service handles this. 
+
+- [r.addb.stats-service.query] Implementation of stats service handles this. 
+
+- [r.addb.stats-service.single-instance] Implementation of stats service handles this.
+
+ Dependencies
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:169]: "Dependencies" has the same level with "Functional Specification" and "Logical Specification".

File: doc/ADDB_Monitoring.rst [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+- [r.addb.record.definition] ADDB monitor can define new addb record. 
+
+- [r.addb.record.definition.extensible] 
+
+- [r.addb.post] ADDB monitor can post addb records. 
+
+- [r.addb.post.non-blocking] Decrease performance impact of ADDB monitoring.
+
+ Use Cases
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:196]: "Use Cases" has the same level with "Functional Specification" and "Logical Specification".

File: doc/ADDB_Monitoring.rst [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+#. Update the state with the information in these latest addb records.
+
+**Query for some state information to the stats service**
+
+#. Construct & send a request fop for specific or complete state information to the stats service & wait for reply. 
+
+#. Stats service checks for requesting information, gathers it in reply fop & sends it back to the node from where request was initiated.
+
+ Failures
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:245]: "Failures" has the same level with "Functional Specification" and "Logical Specification".

Comment (huanghua78) [reply doc/ADDB_Monitoring.rst:245]: Please refer to the original document about these titles.

File: doc/ADDB_Monitoring.rst [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+- A failure to construct new state on the node where the stats service runs would return the previous state to the node that requested this state information during this duration.
+
+- Exceptional conditions are reported to local HA component using a fop, a failure of receiving a fop by local HA component can happen, this would mean that some exceptional conditions can go unnoticed by local HA component. This type of failure is ignored.
+
+ Rationale
+============
+
+The existing ADDB implementation and the newly developed tracing subsystem contributed greatly to the requirement to use C macro interfaces with compile time validation. 
+
+ Compatibility
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:259]: Please conform with the original organization of titles in different levels.

File: doc/ADDB_Monitoring.rst [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+- Inform external world (HA & cluster management tools) about exceptional conditions like failures, overload, etc.
+
+***************
+ Definitions
+***************
+
+ADDB Analysis and Diagnostic Data Base. See [0] for additional details. 
+
+ADDB base record type [Text Wrapping Break]The architecture defines “event”, “data point” and “counter” type ADDB record categories.  See [0] for additional details.  In this HLD we will use the word “exception” instead of “event”.  We’ll use the term “event” to describe any derivation of the base record types.
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:19]: Why is there a "[Text Wrapping Break]"?

File: doc/ADDB_Monitoring.rst [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+- To support online statistics reporting, like similar to df, vmstat, top. This is needed both globally (summarized over all nodes) and locally. 
+
+- Inform external world (HA & cluster management tools) about exceptional conditions like failures, overload, etc.
+
+***************
+ Definitions
+***************
+
+ADDB Analysis and Diagnostic Data Base. See [0] for additional details.
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:17]: The "[0]" reference is not available in this doc.

File: doc/ADDB_Monitoring.rst [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+- [r.addb.monitor.stats-service.single-instance] There would be only one instance of the stats service in the cluster/system.
+
+**************************
+ Functional Specification
+**************************
+
+There are two APIs that are globally visible, they are to add/delete monitors.
+
+::
```
Comment (huanghua78) [inline doc/ADDB_Monitoring.rst:61]: Please try to use:

>  ```C
> xxx code here
> ```
to insert some code

File: doc/ADDB_Monitoring.rst (hunk 5 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst]: Incorporated.

File: doc/ADDB_Monitoring.rst (hunk 4 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst:249]: Incorporated.

File: doc/ADDB_Monitoring.rst (hunk 3 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst:199]: Incorporated.

File: doc/ADDB_Monitoring.rst (hunk 2 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst:171]: Incorporated.

File: doc/ADDB_Monitoring.rst (hunk 1 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst]: Incorporated.

File: doc/ADDB_Monitoring.rst (hunk 6 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst]: Incorporated

File: doc/ADDB_Monitoring.rst (hunk 7 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst]: Incorporated.

**Review by mukul-seagate11 — APPROVED** [review]
OK

File: doc/ADDB_Monitoring.rst (hunk 8 above)
Comment (VenkyOS) [reply doc/ADDB_Monitoring.rst]: Incorporated.

**Review by huanghua78 — APPROVED** [review]

**Pull request merged**